mocktracer/src/dynamic_load.cpp

```diff
@@ -4,29 +4,34 @@
 #include <cstring>
 #include <exception>
 
-static int OpenTracingMakeTracerFactoryFct(const char* opentracing_version,
-                                 const void** error_category,
-                                 void** tracer_factory) {
+static int OpenTracingMakeTracerFactoryFct(const char* /*opentracing_version*/,
+                                           const char* opentracing_abi_version,
+                                           const void** error_category,
+                                           void* error_message,
+                                           void** tracer_factory) try {
   if (error_category == nullptr || tracer_factory == nullptr) {
     fprintf(stderr,
             "`error_category` and `tracer_factory` must be non-null.\n");
     std::terminate();
   }
 
-  if (std::strcmp(opentracing_version, OPENTRACING_VERSION) != 0) {
+  if (std::strcmp(opentracing_abi_version, OPENTRACING_ABI_VERSION) != 0) {
     *error_category =
         static_cast<const void*>(&opentracing::dynamic_load_error_category());
+    auto& message = *static_cast<std::string*>(error_message);
+    message =
+        "incompatible OpenTracing ABI versions; "
+        "expected " OPENTRACING_ABI_VERSION " but got ";
+    message.append(opentracing_abi_version);
     return opentracing::incompatible_library_versions_error.value();
   }
 
-  *tracer_factory =
-      new (std::nothrow) opentracing::mocktracer::MockTracerFactory{};
-  if (*tracer_factory == nullptr) {
-    *error_category = static_cast<const void*>(&std::generic_category());
-    return static_cast<int>(std::errc::not_enough_memory);
-  }
+  *tracer_factory = new opentracing::mocktracer::MockTracerFactory{};
 
   return 0;
+} catch (const std::bad_alloc&) {
```

Comment:
Out of curiosity, why'd you change to `try`/`catch`? If you expect other types of exception, make sure to catch `std::exception` or `catch (...)`.

Reply:
I added the try/catch because it assigns to the string `error_message`, which can allocate memory. `std::bad_alloc` should be the only exception that could be thrown.